File: Java Solution/Defend.java
import java.io.*;
import java.nio.file.*;
import java.security.*;
import java.util.*;

/**
 * Defend Your Code - Java Implementation
 * 
 * This program securely collects user input, validates it, processes files,
 * and handles password security using SHA-256 with salting. It ensures robust
 * error handling, prevents crashes, and logs errors properly.
 * 
 * Features:
 * - Validates user names (only letters, hyphen, and apostrophe allowed)
 * - Ensures integer input is within the 4-byte integer range
 * - Implements password security with hashing and salting
 * - Reads an input file and writes required data to an output file
 * - Logs errors in an error file
 * - Ensures program does not crash due to invalid input
 * 
 * @author Lwazi Murad Sophenith 
 */
public class Defend {
    private static final int MAX_NAME_LENGTH = 50;
    private static final String NAME_REGEX = "^[A-Za-z\\-']+$";
    private static final int INT_MIN = Integer.MIN_VALUE;
    private static final int INT_MAX = Integer.MAX_VALUE;
    private static final String PASSWORD_LOG_FILE = "password.txt";
    private static final String ERROR_LOG_FILE = "error.txt";

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        PrintWriter errorLog = null;
        
        try {
            errorLog = new PrintWriter(new FileWriter(ERROR_LOG_FILE, true));

            // Read user input: First name, Last name, Integers
            String firstName = readName(scanner, "first");
            String lastName = readName(scanner, "last");

            int firstInt = readInt(scanner, "first");
            int secondInt = readInt(scanner, "second");
            long sum = safeAdd(firstInt, secondInt);
            long product = safeMultiply(firstInt, secondInt);

            // Read input and output file names
            String inputFileName = readFileName(scanner, "input");
            String outputFileName = readFileName(scanner, "output");

            while (inputFileName.equals(outputFileName)) {
                System.out.println("Input and output files must be different.");
                outputFileName = readFileName(scanner, "output");
            }

            // Handle password security
            byte[] salt = generateSalt();
            String hashedPassword = "";

            while (true) {
                System.out.print("Enter a password (min 8 chars, 1 uppercase, 1 lowercase, 1 digit, 1 special character): ");
                String password = scanner.nextLine();
                if (!isValidPassword(password)) {
                    System.out.println("Password does not meet requirements. Try again.");
                    continue;
                }

                System.out.print("Re-enter your password for verification: ");
                String passwordVerify = scanner.nextLine();
                if (!password.equals(passwordVerify)) {
                    System.out.println("Passwords do not match. Try again.");
                    continue;
                }

                hashedPassword = hashPassword(password, salt);
                savePassword(hashedPassword, salt);

                System.out.print("Enter your password once more to verify: ");
                String finalPassword = scanner.nextLine();
                if (hashPassword(finalPassword, salt).equals(hashedPassword)) {
                    System.out.println("Password verified successfully.");
                    break;
                } else {
                    System.out.println("Final Password verification failed. Try again.");
                }
            }

            // Read input file contents
            String inputFileContents = "";
            try {
                inputFileContents = new String(Files.readAllBytes(Paths.get(inputFileName)));
            } catch (IOException e) {
                logError(errorLog, "Error reading input file: " + e.getMessage());
                inputFileContents = "Error reading file";
            }

            // Write to output file
            try (PrintWriter writer = new PrintWriter(new FileWriter(outputFileName))) {
                writer.println("First Name: " + firstName);
                writer.println("Last Name: " + lastName);
                writer.println("First Integer: " + firstInt);
                writer.println("Second Integer: " + secondInt);
                writer.println("Sum: " + sum);
                writer.println("Product: " + product);
                writer.println("Input File Name: " + inputFileName);
                writer.println("Input File Contents:\n" + inputFileContents);
                System.out.println("Data successfully written to " + outputFileName);
            } catch (IOException e) {
                logError(errorLog, "Error writing output file: " + e.getMessage());
            }

        } catch (Exception e) {
            logError(errorLog, "An error occurred: " + e.getMessage());
        } finally {
            if (errorLog != null) errorLog.close();
            scanner.close();
        }
    }

    /**
     * Logs an error message to the error log file.
     */
    private static void logError(PrintWriter errorLog, String message) {
        if (errorLog != null) {
            errorLog.println(new Date() + ": " + message);
            errorLog.flush();
        }
    }

    /**
     * Reads and validates a name (first or last).
     */
    private static String readName(Scanner scanner, String type) {
        while (true) {
            System.out.print("Enter your " + type + " name (letters, hyphen (-), apostrophe (') allowed, max " + MAX_NAME_LENGTH + " characters): ");
            String name = scanner.nextLine();
            if (name.length() > 0 && name.length() <= MAX_NAME_LENGTH && name.matches(NAME_REGEX)) {
                return name;
            }
            System.out.println("Invalid name. Please enter again.");
        }
    }

    /**
     * Reads and validates an integer.
     */
    private static int readInt(Scanner scanner, String order) {
        while (true) {
            System.out.print("Enter the " + order + " integer (range: " + INT_MIN + " to " + INT_MAX + "): ");
            try {
                return Integer.parseInt(scanner.nextLine());
            } catch (NumberFormatException e) {
                System.out.println("Invalid integer input. Please enter a valid 4-byte integer.");
            }
        }
    }

    /**
     * Reads a valid file name and ensures input files exist.
     */
    private static String readFileName(Scanner scanner, String type) {
        while (true) {
            System.out.print("Enter the " + type + " file name (e.g., file.txt): ");
            String fileName = scanner.nextLine();
            if (!fileName.isEmpty()) {
                if (type.equalsIgnoreCase("input")) {
                    if (new File(fileName).exists()) return fileName;
                    System.out.println("File does not exist. Try again.");
                } else {
                    return fileName;
                }
            } else {
                System.out.println("File name cannot be empty.");
            }
        }
    }

    /**
     * Validates a password according to complexity requirements.
     */
    private static boolean isValidPassword(String password) {
        if (password.length() < 8) return false;
        return password.matches("^(?=.*[A-Z])(?=.*[a-z])(?=.*\\d)(?=.*[!@#$%^&*()_+=<>?/{}~|]).+$");
    }

    /**
     * Generates a secure random salt.
     */
    private static byte[] generateSalt() {
        byte[] salt = new byte[16];
        new SecureRandom().nextBytes(salt);
        return salt;
    }

    /**
     * Hashes a password using SHA-256 and a provided salt.
     */
    private static String hashPassword(String password, byte[] salt) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            md.update(salt);
            byte[] hashed = md.digest(password.getBytes("UTF-8"));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hashed) hexString.append(String.format("%02x", b));
            return hexString.toString();
        } catch (Exception e) {
            return "";
        }
    }

    /**
     * Saves the hashed password and salt to a file.
     */
    private static void savePassword(String hashedPassword, byte[] salt) {
        try (PrintWriter pw = new PrintWriter(new FileWriter(PASSWORD_LOG_FILE))) {
            pw.println(Base64.getEncoder().encodeToString(salt));
            pw.println(hashedPassword);
        } catch (IOException e) {
            System.out.println("Error saving password.");
        }
    }

    private static long safeAdd(int a, int b) {
        return (long) a + (long) b;
    }

    private static long safeMultiply(int a, int b) {
        return (long) a * (long) b;
    }
}

File: Python Solution/Defend.py
import re
import os
import sys
import hashlib
import secrets
from datetime import datetime

MAX_NAME_LENGTH = 50
NAME_REGEX = r"^[A-Za-z\-']+$"
INT_MIN = -2147483648
INT_MAX = 2147483647
PASSWORD_LOG_FILE = "password.txt"
ERROR_LOG_FILE = "error.txt"

def log_error(message):
    "Log an error message with timestamp to the error log file"
    with open(ERROR_LOG_FILE, 'a') as error_log:
        error_log.write(f"{datetime.now()}: {message}\n")

def read_name(name_type):
    "Read and validate a first or last name."
    while True:
        name = input(f"Enter your {name_type} name (only letters, hyphen (-), apostrophe (') allowed, max {MAX_NAME_LENGTH} characters): ")
        if not name or len(name) > MAX_NAME_LENGTH:
            print(f"Name must be between 1 and {MAX_NAME_LENGTH} characters")
            continue
        if not re.match(NAME_REGEX, name):
            print("Name can only contain letters, hyphen (-), and apostrophe (').")
            continue
        return name
def read_int(order):
    """Read and validate an integer."""
    while True:
        try:
            value_str = input(f"Enter the {order} integer (range: {INT_MIN} to {INT_MAX}): ")
            value = int(value_str)
            if value < INT_MIN or value > INT_MAX:
                print(f"Integer must be within the range {INT_MIN} to {INT_MAX}.")
                continue
            return value
        except ValueError:
            print("Invalid integer input. Please enter a valid 4-byte integer.")

def read_file_name(file_type):
    """Read and validate a file name."""
    while True:
        file_name = input(f"Enter the {file_type} file name (e.g. file.txt): ")
        if not file_name:
            print("File name cannot be empty.")
            continue
        if file_type.lower() == "input":
            if not os.path.isfile(file_name):
                print("File does not exist or is not a file. Please try again.")
                continue
        return file_name
def is_valid_password(password):
    """Validate password meets requirements."""
    if len(password) < 8:
        return False
    
    has_upper = any(c.isupper() for c in password)
    has_lower = any(c.islower() for c in password)
    has_digit = any(c.isdigit() for c in password)
    has_special = any(c in "!@#$%^&*()-_+=<>?/{}~|" for c in password)
    
    return has_upper and has_lower and has_digit and has_special
def generate_salt():
    """Generate a random 16-byte salt."""
    return secrets.token_bytes(16)
def hash_password(password, salt):
    """Hash a password using SHA-256."""
    hasher = hashlib.sha256()
    hasher.update(salt)
    hasher.update(password.encode('utf-8'))
    return hasher.hexdigest()

def save_password(hashed_password, salt):
    """Save the salt and hashed password to a file."""
    with open(PASSWORD_LOG_FILE, "w") as pw_file:
        pw_file.write(salt.hex() + "\n")
        pw_file.write(hashed_password + "\n")

def safe_add(a, b):
    """Safely add two integers without overflow."""
    return a + b

def safe_multiply(a, b):
    """Safely multiply two integers without overflow."""
    return a * b

def main():
    try:
        # Name Input
        first_name = read_name("first")
        last_name = read_name("last")

        # Integer Input
        first_int = read_int("first")
        second_int = read_int("second")
        sum_result = safe_add(first_int, second_int)
        product_result = safe_multiply(first_int, second_int)

        # File Name Input
        input_file_name = read_file_name("input")
        output_file_name = read_file_name("output")
        
        # Make sure input and output files are different
        while input_file_name == output_file_name:
            print("Input and output files must be different.")
            output_file_name = read_file_name("output")

        # Password Input and Verification
        salt = generate_salt()
        hashed_password = ""

        while True:
            password = input("Enter a password (min 8 characters, must include at least one uppercase letter, one lowercase letter, one digit, and one special character): ")
            if not is_valid_password(password):
                print("Password does not meet requirements. Please try again.")
                continue
            
            password_verify = input("Re-enter your password for verification: ")
            if password != password_verify:
                print("Passwords do not match. Please try again.")
                continue
            
            hashed_password = hash_password(password, salt)
            save_password(hashed_password, salt)

            # Final Verification
            final_password = input("Enter your password once more to verify: ")
            final_hash = hash_password(final_password, salt)
            if final_hash == hashed_password:
                print("Password verified successfully.")
                break
            else:
                print("Final Password verification failed. Please try again.")

        # Read Input File
        try:
            with open(input_file_name, 'r') as input_file:
                input_file_contents = input_file.read()
        except Exception as e:
            error_msg = f"Error reading input file: {str(e)}"
            print(error_msg)
            log_error(error_msg)
            input_file_contents = "Error reading file"

        # Write Output File
        try:
            with open(output_file_name, 'w') as output_file:
                output_file.write(f"First Name: {first_name}\n")
                output_file.write(f"Last Name: {last_name}\n")
                output_file.write(f"First Integer: {first_int}\n")
                output_file.write(f"Second Integer: {second_int}\n")
                output_file.write(f"Sum: {sum_result}\n")
                output_file.write(f"Product: {product_result}\n")
                output_file.write(f"Input File Name: {input_file_name}\n")
                output_file.write(f"Input File Contents:\n{input_file_contents}\n")
            print(f"Data successfully written to {output_file_name}")
        except Exception as e:
            error_msg = f"Error writing output file: {str(e)}"
            print(error_msg)
            log_error(error_msg)

    except Exception as e:
        error_msg = f"An error occurred: {str(e)}"
        print(error_msg)
        log_error(error_msg)

if __name__ == "__main__":
    main()